# Сервис для чтения книг по подписке

<div style="border:solid orange 2px; padding: 20px"> <h7 style="color:black; margin-bottom:20px">Привет! Меня зовут Егошин Михаил, я буду проводить ревью на этом спринте. Если ты не против, давай общаться на "ты", так я думаю будет комфортнее и удобнее)
Мои комментарии и замечания далее по тексту помечены различными цветами:
    
---    
<div class="alert alert-block alert-danger">
Критическое замечание, которые следует исправить.
</div> 
    
---
    
<div class="alert alert-block alert-warning">
Замечание\совет на будущее.
</div>
    
---
    
<div class="alert alert-block alert-success">
Когда всё сделано правильно.
</div>
    
---

Для меня главное, не просто указать на ошибки, а чтобы у тебя, в первую очередь, было понимание как это делается, подсказать в каких то моментах. Поэтому если возникнут вопросы по проекту или вопросы общего характера, то не стесняйся - задавай!) Для того, чтобы мне было легче их найти, то выделяй их пожалуйста с помощью цвета, например вот так - <font color=orange>"вопрос, который не останется незамеченным"</font>
    
Удачи!
    
p.s.: не удаляй мои замечания, если предстоит что то доработать в проекте.</h7>

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Итог ревью - ты проделала хорошую работу. Почти все запросы корректные и дают правильный результат. Но есть одно задание, которое тебе предстоит доработать, чтобы я смог принять твой проект. Если возникнут вопросы по моим комментариям, пиши здесь - обязательно отвечу. Буду ждать!
</div>

<font color=orange>Михаил, привет! <a href='#4task'>Четвёртую задачку</a> поправила </font>

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Привет) Все замечания учтены. Успехов в дальнейшем обучении!
</div>

**Задача:**  
Проанализировать базу данных сервиса

**Исходные данные:**  
БД с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string) 

In [3]:
query = '''SELECT * 
FROM books 
LIMIT 5
'''
df = pd.io.sql.read_sql(query, con = engine) 
df

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

В базе содержится информация о тысяче книг

In [4]:
query = '''SELECT * 
FROM authors
LIMIT 5
'''
df = pd.io.sql.read_sql(query, con = engine) 
df

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
query = '''SELECT * 
FROM reviews
LIMIT 5
'''
df = pd.io.sql.read_sql(query, con = engine) 
df

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [6]:
query = '''SELECT * 
FROM ratings
LIMIT 5
'''
df = pd.io.sql.read_sql(query, con = engine) 
df

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
query = '''SELECT * 
FROM publishers
LIMIT 5
'''
df = pd.io.sql.read_sql(query, con = engine) 
df

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<div class="alert alert-block alert-warning">
<b>Комментарии от ревьюера:</b>

Здесь всё хорошо. Ты корректно выгрузила данные и посмотрела их содержимое. Не помешало бы вывести еще информацию с помощью info(). Так у нас будет информация хотя бы о размерности датасетов
</div>

*Посчитайте, сколько книг вышло после 1 января 2000 года*

In [8]:
query1 = '''SELECT COUNT(DISTINCT book_id) 
FROM books
WHERE publication_date > '2000-01-01'
'''
df1 = pd.io.sql.read_sql(query1, con = engine) 

In [9]:
df1

count
0    819

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Всё верно. Это правильный ответ (правда distinct здесь ни на что не влияет, т.к. book_id является уникальным идентификатором, а джойнов никаких нет в запросе)
</div>

После 1 января 2000 года вышло 819 книг - бОльшая часть из всех книг в таблице. То есть наш сервис ориентирован на современную литературу

*Для каждой книги посчитайте количество обзоров и среднюю оценку*

In [10]:
query2 = '''SELECT b.book_id, n_reviews,avg_rating
FROM books b
LEFT JOIN (
SELECT book_id, COUNT(review_id) n_reviews
FROM reviews
GROUP BY book_id) r ON r.book_id=b.book_id
LEFT JOIN (
SELECT book_id, AVG(rating) avg_rating
FROM ratings
GROUP BY book_id) ra ON ra.book_id=b.book_id
ORDER BY avg_rating DESC
'''
df2 = pd.io.sql.read_sql(query2, con = engine) 

In [11]:
df2

book_id  n_reviews  avg_rating
0        518        2.0        5.00
1        732        2.0        5.00
2        347        2.0        5.00
3        610        2.0        5.00
4        330        1.0        5.00
..       ...        ...         ...
995      915        3.0        2.25
996      371        2.0        2.00
997      316        2.0        2.00
998      202        3.0        2.00
999      303        2.0        1.50

[1000 rows x 3 columns]

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Здорово. Это один из самых коварных запросов, как показывает практика, но ты справилась успешно. Молодец!)
</div>

У книг с самыми высокими рейтингами необязательно много текстовых обзоров

*Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры*

In [12]:
query3 = '''SELECT publisher,
COUNT(book_id) n_books
FROM books b
INNER JOIN publishers p ON p.publisher_id=b.publisher_id
WHERE num_pages>50
GROUP BY publisher
ORDER BY n_books DESC 
LIMIT 1
'''
df3 = pd.io.sql.read_sql(query3, con = engine) 

In [13]:
df3

publisher  n_books
0  Penguin Books       42

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Согласен с результатом. Действительно данное издательство является лидером по выпущенным книгам
</div>

Больше всего книг вышло у издательства Penguin Books (одно из крупнейших мировых издательств)

<a id='4task'></a>

*Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками*

In [37]:
query4 = '''SELECT author,
AVG(rating) avg_rating
FROM ratings r
INNER JOIN (SELECT b.book_id, author_id 
FROM ratings r
INNER JOIN books b ON b.book_id=r.book_id
GROUP BY b.book_id, author_id 
HAVING COUNT(rating_id)>=50
) b ON b.book_id=r.book_id
INNER JOIN authors a ON a.author_id=b.author_id
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
'''
df4 = pd.io.sql.read_sql(query4, con = engine) 
df4

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

<div class="alert alert-block alert-danger">
<b>Комментарии от ревьюера:</b>

Не совсем верный ответ. Ты подсчитавешь ```HAVING COUNT(rating_id) >= 50``` по авторам, а не по книгам (т.е. выводишь авторов у которых 50 и более оценок, а надо книги). Предлагаю первым подзапросом отфильтровать книги по условию задачи, а потом уже искать кто является их автором и рассчитывать показатели
</div>

Победила Джоан Роулинг :)

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера v2:</b>

Да, теперь результат корректный
</div>

*Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок*

In [16]:
query5 = '''
SELECT 
AVG(n_review) AVG_review
FROM
(SELECT username, 
COUNT(rating_id) n_rating
FROM ratings
GROUP BY username
HAVING COUNT(rating_id)>50) r
LEFT JOIN (SELECT username,
COUNT(review_id) as n_review
FROM reviews
GROUP BY username) re ON re.username=r.username

'''
df5 = pd.io.sql.read_sql(query5, con = engine) 

In [17]:
df5

avg_review
0   24.333333

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

И это .. правильный ответ! Ты корректно отобрала пользователей и рассчитала их показатели.

В итоге, 4\5 с первой попытки - это достойный результат)
</div>

В среднем пользователи с более чем 50 оценками оставляют примерно 24 обзора. На первый взгляд это неплохой результат. Почти в половине случаев такие пользователи наряду с оценкой оставляют текстовый комментарий

**Основные выводы:**  
Сервис ориентирован на современную массовую литературу: большинство книг вышло после 2000 года, в лидерах по рейтингам - авторы самых массовых книг.  
Пользователи, которые ставят оценки примерно в половине случаев оствляют текстовый обзор. Можно дополнительно стимулировать пользователей к тому, чтобы они делились мнениями и обсуждали прочитанное, так как это будет привлекать новых пользователей и они будут больше задерживаться в сервисе не только для чтения, но и для обсуждения прочитанного

<div class="alert alert-block alert-success">
<b>Комментарии от ревьюера:</b>

Здорово, что ты не просто вывела полученные результаты, но и дала грамотные рекомендации бизнесу! Это важное качество для аналитика. Развивай его и дальше)
</div>